### ISSUE TITLE
dnpNMR.integrate replaces the 'proc' spectrum

### ISSUE THREAD

**Issue by thcasey3** [DNPLab/DNPLab#4]
dnpNMR.integrate changes the nature of the data, it is not an additional processing step. It should add the new single point, or array for 2D data, as a separate entry and 'proc' should still be the processed spectrum. For example, the dnpFit module doesn't replace 'proc', just uses 'proc' to make and add ['fit'], .attrs['t1'], and .attrs['t1_stdd']. This should be the way the integrate function works as well

**Comment by thorstenmaly** [comment]
I think it is actually correct that proc is replaced with the results of the processing step. To keep the 2D array copy the data to a new key before integrating.
This is a broader issue that requires better documentation.

**Comment by thcasey3** [comment]
The origin of my concern was using sliders in the GUI or iteration to optimize the integration width or center. For iterative processes the difference in code would be something like:
```
workspace.copy("proc", "fft") 
for k in range(...): 
   workspace.copy("fft", "proc")
   dnplab.dnpNMR.integrate(workspace, {"integration_width : k})
   integral[k] = workspace["proc"].values
...code to find best k...
workspace.copy("fft", "proc")
best_integral = dnplab.dnpNMR.integrate(workspace, {"integration_width : best_k})
...code to plot NMR spectrum...
```
where it could be something like:
```
for k in range(...): 
   dnplab.dnpNMR.integrate(workspace, {"integration_width : k})
   integral[k] = workspace["integral"].values
...code to find best k...
best_integral = dnplab.dnpNMR.integrate(workspace, {"integration_width : best_k})
...code to plot NMR spectrum...
```
no copying necessary. 

Currently, for a slider the same three copy statements are necessary but they would need to be outside the function that operates when the slider is moved, one in any code before it and the other in any function that follows the slider movement. This can become very redundant when there are multiple functions that can run in any order. It ends up easier to keep track of everything by just making copies of the whole workspace at the beginning of every function and operating only on the local copies. Again, very redundant usage of copying. I would maybe understand this if this type of syntax ```new_workspace = dnplab.dnpNMR.integrate(workspace, {"integration_width : k})``` didn't alter "proc" in the workspace and I could make new workspaces to preserve the original, but this also overwrites "proc" in the original workspace, and even if it didn't, again, always a new workspace just for an integration. What if, for example, ```value.sqrt()``` AND ```new_value = sqrt(value)``` both wiped out ```value```? 

Effectively, having integrate add an extra entry to the workspace OR requiring the user to manually add another entry first starts with one workspace entry and ends with two entries, it just requires more lines of code. While I agree that we shouldn't add every spectrum that is the result of a processing step to the workspace as a separate entry, a function that takes the spectrum and creates something that is not a spectrum, like a fit or integration, warrants a new entry. 

If it is going to stay this way the documentation should explain clearly where and why a user's code or workflow would benefit from this behavior, or where their code or workflow would be complicated by not having this behavior. Maybe a brief counter example can be given here?

**Comment by jmfranck** [comment]
@thcasey3 : first, I should point out that @ylin00 has posted a commit here.

The best way I can "chime in" as requested here is to give a different perspective:  The way you would do this from a more pyspecdata-like codebase would be just:
```python
for j,int_width in enumerate(...):
   int_range = tuple(int_center+r_[-0.5,0.5]*int_width)
   integral['width',j] = data['t2':int_range].integrate('t')
```

The ``data['t2':sometuple]`` you can think of as a new syntax that we provide that lets you slice by the coordinates, rather than the index numbers (you can still do ``data['t2',0:100]`` to get, e.g. the first 100 elements -- note comma vs. colon).  This is the only such weird addition that we make, b/c it's pretty powerful and leads to compact code -- e.g. `data['t2':(0,0.5)]` directly gives you the first 0.5 seconds of an FID.

Directly addressing your comment, in the above code, there is no need to specify an additional copy, because there's an implied need to copy when you take the integral or sum of a slice.

Philosophically, I like doing things as I show above.  Keeping the best of both worlds, I would like to incorporate @tkeller12 's code where `integral` would contain a record where it would know that it was created from data, sliced, and integrated -- this is pretty straighforward to incorporate, I think, and importantly wouldn't change the actual snippet above in any way (work would be done behind the scenes).

Sorry for the long-winded and somewhat orthogonal answer, but I hope it's helpful in the long run.

**Comment by thcasey3** [comment]
@jmfranck this looks nice, however I would argue that the bulk of users that are looking for a new solution to NMR data problems wouldn't be able to (without some extra effort) or even want to learn how to construct code like this.  This does not simplify their efforts. Those who would have no problem constructing this type of code aren't looking for a new solution in the first place, and if they were they would only switch from their current method if the simplification we offer is enticing. If interacting with DNPLab is at the same level of complexity as what they are currently doing they aren't going to bother re-writing all their code. This is why I think we are on the right track but only a few little steps off from making something uniquely simple, like this:
```
workspace = dnplab.dnpImport(path, type="topspin")
dnplab.dnpNMR.remove_offset(workspace, {})
dnplab.dnpNMR.window(workspace, {"linewidth": 5})
dnplab.dnpNMR.fourier_transform(workspace, {"zero_fill_factor": 2})
dnplab.dnpNMR.align(workspace, {})
dnplab.dnpNMR.integrate(workspace, {"integrate_center" :  0, "integrate_width" : 100})
dnplab.dnpFit(workspace, type="invrecov")
```
Only seven lines of "code" to completely process, integrate, and fit. No python expertise necessary, and quick access to everything at any point with simply:
```
fid = workspace["raw"].values
time = workspace["raw"].coords
fft = workspace["proc"].values
frequency = workspace["proc"].coords
integrals = workspace["integral"].values
integral_axis = workspace["integral"].coords
fit = workspace["fit"].values
fit_axis = workspace["fit"].coords
T1 = workspace["fit"].attrs["T1"]
```
In my opinion, this simplicity would entice anyone, expert or not. pyspecdata offers a wide range of expert level functionality, for higher level users we can refer them to pyspecdata rather than re-invent the wheel in DNPLab.

**Comment by ylin00** [comment]
> The origin of my concern was using sliders in the GUI or iteration to optimize the integration width or center. For iterative processes the difference in code would be something like:
> 
> ```
> workspace.copy("proc", "fft") 
> for k in range(...): 
>    workspace.copy("fft", "proc")
>    dnplab.dnpNMR.integrate(workspace, {"integration_width : k})
>    integral[k] = workspace["proc"].values
> ...code to find best k...
> workspace.copy("fft", "proc")
> best_integral = dnplab.dnpNMR.integrate(workspace, {"integration_width : best_k})
> ...code to plot NMR spectrum...
> ```
> 
> where it could be something like:
> 
> ```
> for k in range(...): 
>    dnplab.dnpNMR.integrate(workspace, {"integration_width : k})
>    integral[k] = workspace["integral"].values
> ...code to find best k...
> best_integral = dnplab.dnpNMR.integrate(workspace, {"integration_width : best_k})
> ...code to plot NMR spectrum...
> ```
> 
> no copying necessary.


dnpNMR.integrate has two kinds of behaviors: if passing a Workspace, then it will change the 'proc' value; if passing a DNPData object, it will return the integral. See [test_dnpNMR.py#L26 ](https://github.com/DNPLab/DNPLab/blob/97cb45ca0310dff0884fe93af3bf471e25f2f0dd/unittests/test_dnpNMR.py#L26)

Maybe the workspace model is not suitable for multiple trials of integration?

**Comment by jmfranck** [comment]
@thcasey3 : yes, but what you are pointing to is a level of code up from the question that I thought was being asked.  The type of code that I'm showing could be used inside functions that perform alignment (or integration, now that I understand the point of your function a bit better.)

I honestly see a few areas for improvement the code that you show:

- It is very verbose.  I don't understand the need for a dictionary-like object containing "proc" "integrals" and "raw" -- why would these not just be separate variables?  For that matter, why do we even need separate variables, especially if we're going to keep a record of why the data was processed?  Why not just process the data linearly, step by step, using the minimal amount of code to update it at each step?  This is something worth considering.
- Why are you passing dictionaries rather than using keyword arguments?
- Yes, we can "tear apart" the core object, as you are doing at the end, but the point is that the user *should not need to do this* -- this is why we need basic examples to throw around.

"Simplicity" is in many cases an issue of familiarity, so we should be careful.  In particular, we do not want to carry over the clunkiness of Matlab when we have a fully object-oriented programming environment at our disposal.  The types of capabilities I show above will *massively* simplify users efforts if we can provide them with some basic examples of how to use such code.

**Comment by thcasey3** [comment]
To your points:

1) that is seven full processing/fitting steps in seven lines that works on 1D or 2D data without any extra arguments, that seems to be minimally verbose to me. This gets at the heart of my concern here. Why do we want things to go away unless we make a copy of them? Are we trying to reduce clutter? Save memory? I don't get it. It will be very common for people to want to do something with the spectrum after it is integrated, why should it be gone unless I copy it? I use the same type of example again, what if you did ```new_value = numpy.divide(value, 2)``` and you lose ```value``` unless you make a copy of it first? Why can't ```new_value``` be something new and have ```value``` left alone? 

2) this is the way it is currently setup. I think we should change it to the style of all major packages, i.e. keyword arguments like ```dnplab.dnpNMR.integrate(workspace,  center=0, width=100)```

3) why is it considered "tearing it apart" to have access the results of the functions? People may want to plot it in their own way, use functions outside of dnplab, etc. and it should be easy to access each component of the data as its own variable if needed. Of course, interaction with the object can be different if you don't need to access the individual components separately but it should be an easy option.

**Comment by thcasey3** [comment]
And to be clear, I don't think this should be true for all operations. For windowing, for example, it will be very rare for a user to be interested in having the FID with AND without windowing, in that case they should use our copy function. However, it will be VERY common for the user want the spectrum AND the values of the integrals, just like it will be VERY common for the user to want the results of a fit AND the subject of the fit.

**Comment by jmfranck** [comment]
I'll cherry-pick a bit here, since we ended up addressing a bit in the meeting:

I prefer the philosophy of progressively processing the data where we think of progressively manipulating the same object -- like you start with a block of rock, and make it into a sculpture.  First we FT it, then we might slice out a particular frequency range, then we might apply a phase correction to it, etc.  It does end up being the most compact to just progressively manipulate the same variable.  If you want to make a copy of the data at any point, you're free to do that.  For example, once I FT the FID, I typically have no need to keep the FID hanging around.  The other benefit of this strategy is that in-place operations actually involve less typing.

I can best explain by example.  Trying to "package" as few functions as possible, I would get from the raw data to the integrals like so.  Note that this is a bit different than what you're doing above, because I'm going for code that is *explicit in what it's doing* and only farms off to other functions when there are genuinely a couple different options to choose between (to be clear, I'm not necessarily advocating that the user needs to know all their Fourier theorems off the cuff, and I'm not disagreeing with the fact that you've wrapped several of these steps into one on the user end):

```python
d = find_file(....)
center_shift = determine_centerfreq(d)
d *= exp(-1j*2*pi*center_shift)
d.ft('t2', shift=True)
ph0, ph1 = determine_corrections(d)
d *= exp(-1j*2*pi*ph0)
d *= exp(-1j*2*pi*ph1*d.coords['t2'])
integrals = d['t2':(start_freq,stop_freq)].integrate('t2')
```
I think this also illustrates that I mean something slightly different when I say "verbose" vs. "compact".  I'm not talking about how many lines of code I've wrapped up inside a library function, but rather how much typing I need to do to get the code to perform a particular action.  I put a lot of value on that, because it's the amount of time that I or my students need to invest to maintain complete control of my results, while generating them quickly.
